### PR TITLE
[Custom OP]add PD_THROW and PD_CHECK for User Error message

### PR DESCRIPTION
**Pull request by zhwesky2010**
### PR types

New features
### PR changes

OPs
### Describe

用户在自定义OP中使用宏 `PD_CHECK` 、 `PD_THROW`添加报错信息，支持变长参数模板
------

**1. PD_THROW()**
---
``An error occured at /zhouwei25/Paddle/python/paddle/fluid/tests/custom_op/custom_relu_op.cc:82``

**2. PD_THROW("PD_THROW returns ", false, ". DataType of ", 1, " is INT")**
---
``PD_THROW returns false. DataType of 1 is INT at /zhouwei25/Paddle/python/paddle/fluid/tests/custom_op/custom_relu_op.cc:82``

**3. PD_CHECK(a > b)**
---
``Expected a > b, but it's not satisfied at /zhouwei25/Paddle/python/paddle/fluid/tests/custom_op/custom_relu_op.cc:82``

**4. PD_CHECK(c < d, "PD_CHECK returns false ", because c  > d")**
---
``"PD_CHECK returns false, because  c > d  at /zhouwei25/Paddle/python/paddle/fluid/tests/custom_op/custom_relu_op.cc:82``

例如，在Windows10上：
image

### CODE REVIEW

File: paddle/fluid/extension/include/op_meta_info.h [hunk 1]
```diff
@@ -47,6 +47,60 @@ using Tensor = paddle::Tensor;
...
 
+#if !defined(_WIN32)
+#define PD_UNLIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 0))
+#define PD_LIKELY(expr) (__builtin_expect(static_cast<bool>(expr), 1))
+#else
+#define PD_UNLIKELY(expr) (expr)
+#define PD_LIKELY(expr) (expr)
+#endif
+
+struct PD_Exception : public std::exception {
```
Comment (chenwhql) [inline paddle/fluid/extension/include/op_meta_info.h]: I think split Exception、CHECK、THROW to a new header file `exception.h` is better

File: paddle/fluid/extension/include/op_meta_info.h [hunk 2]
```diff
@@ -60,11 +114,25 @@ using Tensor = paddle::Tensor;
...
-  do {                                 \
-    HANDLE_THE_ERROR                   \
-    throw std::runtime_error(err_msg); \
-    END_HANDLE_THE_ERROR               \
+#define PD_CHECK(COND, ...)                                               \
+  do {                                                                    \
+    if (PD_UNLIKELY(!(COND))) {                                           \
+      HANDLE_THE_ERROR                                                    \
+      auto __message__ = ::paddle::ErrorMessage(__VA_ARGS__).to_string(); \
+      throw ::paddle::PD_Exception(__message__, __FILE__, __LINE__,       \
```
Comment (chenwhql) [inline paddle/fluid/extension/include/op_meta_info.h]: 这个file和line的格式能再整理一下吗，现在回导致第一行信息太长，是不是换行之后缩进一下，用括号包起来，比如
```
ExternalError: error message.
  [file:line]
```

File: paddle/fluid/extension/include/op_meta_info.h [hunk 3]
```diff
@@ -60,11 +114,25 @@ using Tensor = paddle::Tensor;
...
+                                   "Expected " #COND                      \
```
Comment (chenwhql) [inline paddle/fluid/extension/include/op_meta_info.h]: 这两个提升我觉得可以没有，因为这个是给用户用的，他自己清楚自己写了什么报错，加了多余的信息，他反而会觉得疑惑
框架里面有，主要是为了方便外部用户反馈过来的时候，我们RD能够比较快找到报错语句

File: paddle/fluid/extension/include/dispatch.h [hunk 4]
```diff
@@ -41,9 +41,8 @@ namespace paddle {
       PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::FLOAT64, double,  \
                            __VA_ARGS__)                                \
       default:                                                         \
-        throw std::runtime_error("function " #NAME                     \
-                                 " not implemented for data type `" +  \
-                                 ::paddle::ToString(__dtype__) + "`"); \
+        PD_THROW("function " #NAME " not implemented for data type `", \
+                 ::paddle::ToString(__dtype__), "`");                  \
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/dispatch.h]: Grammar problem：`is not implemented`

File: paddle/fluid/extension/include/dispatch.h [hunk 5]
```diff
@@ -63,27 +62,25 @@ namespace paddle {
       PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::INT16, int16_t,          \
                            __VA_ARGS__)                                       \
       default:                                                                \
-        throw std::runtime_error("function " #NAME                            \
-                                 " not implemented for data type `" +         \
-                                 ::paddle::ToString(__dtype__) + "`");        \
+        PD_THROW("function " #NAME " not implemented for data type `" +       \
+                 ::paddle::ToString(__dtype__) + "`");                        \
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/dispatch.h]: Same as above

File: paddle/fluid/extension/include/dispatch.h [hunk 6]
```diff
@@ -63,27 +62,25 @@ namespace paddle {
...
+#define PD_DISPATCH_COMPLEX_TYPES(TYPE, NAME, ...)                      \
+  [&] {                                                                 \
+    const auto& __dtype__ = TYPE;                                       \
+    switch (__dtype__) {                                                \
+      PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::COMPLEX64,         \
+                           ::paddle::complex64, __VA_ARGS__)            \
+      PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::COMPLEX128,        \
+                           ::paddle::complex128, __VA_ARGS__)           \
+      default:                                                          \
+        PD_THROW("function " #NAME " not implemented for data type `" + \
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/dispatch.h]: Same as above

File: paddle/fluid/extension/include/dispatch.h [hunk 7]
```diff
@@ -106,31 +103,29 @@ namespace paddle {
       PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::INT16, int16_t,          \
                            __VA_ARGS__)                                       \
       default:                                                                \
-        throw std::runtime_error("function " #NAME                            \
-                                 " not implemented for data type `" +         \
-                                 ::paddle::ToString(__dtype__) + "`");        \
+        PD_THROW("function " #NAME " not implemented for data type `" +       \
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/dispatch.h]: Same as above

File: paddle/fluid/extension/include/dispatch.h [hunk 8]
```diff
@@ -157,9 +152,8 @@ namespace paddle {
       PD_PRIVATE_CASE_TYPE(NAME, ::paddle::DataType::COMPLEX128,              \
                            ::paddle::complex128, __VA_ARGS__)                 \
       default:                                                                \
-        throw std::runtime_error("function " #NAME                            \
-                                 " not implemented for data type `" +         \
-                                 ::paddle::ToString(__dtype__) + "`");        \
+        PD_THROW("function " #NAME " not implemented for data type `" +       \
+                 ::paddle::ToString(__dtype__) + "`");                        \
```
Comment (Aurelius84) [inline paddle/fluid/extension/include/dispatch.h]: Same as above.

File: paddle/fluid/extension/include/op_meta_info.h (hunk 2 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/op_meta_info.h]: Done

File: paddle/fluid/extension/include/op_meta_info.h (hunk 3 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/op_meta_info.h]: 默认的报错信息，用户如果写了就以用户为准

File: paddle/fluid/extension/include/dispatch.h (hunk 4 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/dispatch.h]: Done

File: paddle/fluid/extension/include/dispatch.h (hunk 5 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/dispatch.h]: Done

File: paddle/fluid/extension/include/dispatch.h (hunk 7 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/dispatch.h]: Done

File: paddle/fluid/extension/include/dispatch.h (hunk 8 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/dispatch.h]: Done

File: paddle/fluid/extension/include/op_meta_info.h (hunk 1 above)
Comment (zhwesky2010) [reply paddle/fluid/extension/include/op_meta_info.h]: Done, split to ``exception.h``

**Review by chenwhql — APPROVED** [review]

**Pull request merged**